### ISSUE TITLE
Feature request: detect runtime version and inform user if wrong version is used

### ISSUE THREAD

**Issue by jurkstas** [at67/gigatron-rom#7]
### User story
As a programmer I expect to de warned (or error thrown at me) if I am trying to compile using a wrong runtime version.

**Comment by at67** [comment]
1) Do you mean for example if you have specified a version pragma lower than the functionality that you are trying to use? e.g:
~~~
_codeRomType_ ROMv1
        .
        .
        .
load sprite blah
sprite blah
~~~

2) Or do you mean compiling/running versions higher than the hardware's/emulator's ROM version that you are uploading to?

The runtime handles (2) now by checking the ROM version at code startup and continuously flashing the top left hand corner pixel indefinitely until reset, (as suggested by Marcel).

For (1) currently the programmer is expected to know the feature set supported by each ROM version and setting his ROM version pragma accordingly. I could look into adding error messages for this scenario.

**Comment by jurkstas** [comment]
I mean the runtime pragma `_runtimePath_ "../runtime"`. If this line is missing, build fails. If I copy the runtime folder to my project folder so I would not need to have a weird path (something like `_runtimePath_ "../../at67/gigatron-rom/contrib/at67/gtbas/runtime"` ) and forget to update it, compiled code does not run properly. Hence my suggestion to check if the runtime is the correct one.

Maybe it would make more sense to  have a default runtime (just like the default ROM) that is used unless `_runtimePath_` overrides it?

Regarding ROM, (1) seems reasonable, given documentation. (2) is a nice feature that I was not aware of, thanks.

**Comment by at67** [comment]
I've set the default runtime path to "./runtime", so that if the pragma is missing it will by default search for a runtime folder in the current directory.

But from reading your request a 2nd time, I think you are asking for more than that? Maybe some sort of versioning or ID system that tags different runtime branches and allows you to choose based on that tag?

**Comment by jurkstas** [comment]
I see what I did here. I intended to write this as a user story, but we have not agreed to this format and now it probably looks rude and/or entitled, doesn't it? Sorry about that.

What is your intention with the runtime folder? Is it supposed to be redistributed with source? Should it stay with compiler? Something else? I might just be doing things wrong from the start.

**Comment by at67** [comment]
Nothing of the sort, I have no problem whatsoever with you suggesting features, ideas or additions.

That is a good question, to be honest I haven't really thought it about it in detail; when I came up with the idea of using source code at the linking level I wasn't sure how useful it would be. But it seems to have worked out well, it saved me a lot of time and it's trivial to make changes to the runtime and test them without having to jump through any hoops.

As of now the runtime is similar to the firmware, each version of it is specifically tied to and tightly coupled to a unique version of the compiler, (e.g. it would not be trivial to mix and match different revisions of each), but it seems that now that it has become a lot more stable it is being updated much less often than the compiler source itself.

Right now I envision that the runtime is basically part of the compiler source given how tightly coupled they are, so each compiler version should be shipped with it's own unique runtime. But that doesn't stop anyone from modifying the runtime or making additions etc.

If you can think of ways or features to decouple the compiler and runtime further that allow better modularity or allow more flexibility, then I am keen to explore them!

The only thing I would be wary of is causing mismatches between compiler and runtime and/or confusion for less knowledgeable users and programmers.

**Comment by jurkstas** [comment]
It all makes sense. When I asked about including files I was thinking about code re usability as libraries (in addition to being able to split code into smaller chunks).
> The only thing I would be wary of is causing mismatches between compiler and runtime and/or confusion for less knowledgeable users and programmers.
That's exactly what happened to me, haha. I would suggest coupling compiler and runtime more, not less, from user's perspective.

**Comment by at67** [comment]
The compiler doesn't really have a concept of libraries as a traditional compiler does, (usually through an object code/binary linker); but it does now have the module command which can allow you pull in reusable modules. The linker in this project only understands vCPU source code and is currently only being used for pulling in only bits of the runtime that the compiler actually needs.

The module command should allow you to setup and use reusable code modules at the BASIC source level, have you taken a look at it yet? It's super simple: MODULE "filename" where filename contains an absolute or relative path and the file itself can contain, (theoretically), any BASIC source code whatsoever. The module command can be nested as well, but I haven't put any serious testing into this part of it, i.e. it would probably fall over in a heap if you encountered corner cases like circular dependencies etc; obviously we can add a guard/fence system to it later if needed.

It's trivial how it works, it effectively inserts the module into the main source file at exactly the point that the MODULE command is found; error checking also takes into account the changes and does it's best to print which module the error belongs to, (this needs further testing).

P.S. Because the assembler is effectively built into the compiler, the modules themselves can be straight BASIC code, straight vCPU code or a mixture of both, so there is no limitation in terms of what type of code you can put into modules; also there is no naming requirements, the modules can be named in any way you like, i.e. they don't have to be .i or .h or .m etc.

**Comment by at67** [comment]
I've added a version number to "util.i" within the runtime, i.e. %define RUNTIME_VERSION 100, this version number has to exactly match the compiler's internal runtime version number or the compiler produces an error message. All runtime versions before 100 are by default 0, i.e. they are missing the define so they get assigned a default value of 0. This number will only change when the runtime changes, it will not change if only the compiler changes.

Does this sound like the functionality you were looking for? It was trivial to add and I fixed some code duplication issues and bugs in the define handler along the way.

P.S. I also changed the left hand top flashing pixel error message for incorrect ROM versions to the center pixel, it's much easier to see now and app's like PucMon disable the top few scanlines for extra code space.
